[C10D] Add tests for gather and gather_object with subgroup

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118359

Addresses #118337 somewhat- we probably need to update docs. Let's first
confirm what behavior we want.

Identifies a couple of confusing things
1) 'dst' arg for many collectives is always in 'global' rank regardless
   of whether a subgroup is passed in.  This needs a doc update
2) gather_object has a strong dependency on setting the cuda device;
   could we make that smoother?
3) gather_object also should be happy with an empty list on the dst
   side, imo

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @yf225

wconstab committed:
[C10D] Add tests for gather and gather_object with subgroup

Addresses #118337 somewhat- we probably need to update docs. Let's first
confirm what behavior we want.

Identifies a couple of confusing things
1) 'dst' arg for many collectives is always in 'global' rank regardless
   of whether a subgroup is passed in.  This needs a doc update
2) gather_object has a strong dependency on setting the cuda device;
   could we make that smoother?
3) gather_object also should be happy with an empty list on the dst
   side, imo

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add tests for gather and gather_object with subgroup"


Addresses #118337 somewhat- we probably need to update docs. Let's first
confirm what behavior we want.

Identifies a couple of confusing things
1) 'dst' arg for many collectives is always in 'global' rank regardless
   of whether a subgroup is passed in.  This needs a doc update
2) gather_object has a strong dependency on setting the cuda device;
   could we make that smoother?
3) gather_object also should be happy with an empty list on the dst
   side, imo

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225

[ghstack-poisoned]

wconstab commented:
@kwen2501 wonder if you know off the top of your head whether the same doc change i made for the gather variants applies to the src rank for the scatter variants, and any other ops (send/recv, perhaps)? We need to update them all but I ran out of steam tonight.

fegin commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2369,7 +2369,7 @@ def gather_object(obj, object_gather_list=None, dst=0, group=None):
             collective and will contain the output. Must be ``None`` on non-dst
             ranks. (default is ``None``)
-        dst (int, optional): Destination rank. (default is 0)
+        dst (int, optional): Destination rank on global process group (regardless of 'group' argument). (default is 0)
```
Even if the document is updated, the behavior is very counterintuitive. Should we change to use the group's rank?

wconstab replied:
I prefer the changed behavior.  But i do not know if it is worth the disruption of making a breaking change. worth a discussion

wconstab replied:
if we want to change it, what we might have to do is first add a 'group_rank' arg and add logic to ensure only one of group_rank or rank is passed.  Then deprecate rank.  then finally remove rank.  Or how else?

weifengpy reviewed (approved):
approve since this is more about updating docs and verify current behavior. 
will update docs / unit tests similarly for send/reduce with ``dst`` param

weifengpy commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -3529,6 +3529,61 @@ def test_new_group_local_sync_sanity_check(self):
+            return
+
+        store = c10d.FileStore(self.file_name, self.world_size)
+        torch.distributed.init_process_group(backend="nccl", store=store, rank=self.rank, world_size=self.world_size)
```
on devgpu with 8 gpus (world_size=8), seems only rank 0...3 calls ``torch.distributed.init_process_group`` ? PG does not complain about no init from rank 4...7 ?

wconstab replied:
good point. since I chose to make my test run with exactly 4 gpus for simplicity, i should not rely on self.world_size to report 4 in case someone changes it.  I will update this to pass 4 into init_process_group

weifengpy commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -3529,6 +3529,61 @@ def test_new_group_local_sync_sanity_check(self):
+    @requires_nccl()
+    @skip_if_lt_x_gpu(4)
+    def test_gather_subgroup(self):
+        if self.rank > 3:
```
do we want to set world_size explicitly? right now it's hard coded to 4 from from ``test_c10d_common.AbstractLargeCommTest`` any way

```
@property
    def world_size(self):
        return 4
```

wconstab replied:
i wrote the test in as simple a way as possible, hardcoding lists of ranks e.g. [0, 1] instead of doing things like `ranks = list(range(self.world_size))[:self.world_size/2]`.  It just felt easier to read and even less LOC this way.  It is not that important to have the test run on 8 gpus if they are available.

weifengpy replied:
thanks. it's clear now with ``world_size=4`` in ``init_process_group``

wconstab committed:
Update on "[C10D] Add tests for gather and gather_object with subgroup"


Addresses #118337 somewhat- we probably need to update docs. Let's first
confirm what behavior we want.

Identifies a couple of confusing things
1) 'dst' arg for many collectives is always in 'global' rank regardless
   of whether a subgroup is passed in.  This needs a doc update
2) gather_object has a strong dependency on setting the cuda device;
   could we make that smoother?
3) gather_object also should be happy with an empty list on the dst
   side, imo

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge
